frontend/pages/admin/charities/[charityId]/edit.tsx

```diff
@@ -0,0 +1,91 @@
+import { Box, Container } from '@mui/system';
+import Head from 'next/head';
+import useSWR from 'swr';
+import CharityForm, { charitySchema } from '../../../../components/charities/form/CharityForm';
+import { containerSx } from '../../../../styles/admin/charities/createStyles';
+import { CharityData, CharityFormData, CharityPutData } from '../../../../types/charity';
+import * as Yup from 'yup';
+import api from '../../../../frontendApis';
+import { useRouter } from 'next/router';
+import { Stack, Typography } from '@mui/material';
+import { errorStackSx } from '../../../../styles/admin/charities/editStyles';
+import useAdminLoginCheck from '../../../../hooks/useAdminLogInCheck';
+
+const CharityEdit = () => {
+  useAdminLoginCheck();
+  const router = useRouter();
+
+  // Note that the router query parameters will all be undefined on initial render.
+  const { query } = router;
+  const charityId = query.charityId ? Number(query.charityId) : undefined;
+  const { data: initialValues, error } = useSWR<CharityFormData | undefined>([charityId], getCharityData);
+  const isLoading = !initialValues && !error;
+
+  const handleSubmit = (values: CharityFormData) => {
+    charitySchema
+      .validate(values)
+      .then((values: Yup.InferType<typeof charitySchema>) => {
+        const charityPostData: CharityPutData = {
+          ...values,
+        };
+
+        return api.charities.putCharity(charityId as number, charityPostData);
+      })
+      .then(() => {
+        router.push('/admin/charities');
+      });
+  };
+
+  return (
+    <Box>
+      <Head>
+        <title>Edit Charity</title>
+      </Head>
+
+      <Container sx={containerSx} component="main" maxWidth="sm">
+        {initialValues ? (
+          <CharityForm
+            title="Edit Charity"
+            submitButtonTitle="Edit"
+            initialValues={initialValues}
+            onSubmit={handleSubmit}
+          />
+        ) : (
+          <></>
+        )}
+        {isLoading ? (
+          <Stack>
+            <Typography variant="h1">Loading...</Typography>
+          </Stack>
+        ) : (
+          <></>
+        )}
+        {error ? (
+          <Stack spacing={2} sx={errorStackSx}>
+            <Typography variant="h1">Error</Typography>
+            <Typography variant="h2">That is all we know right now.</Typography>
+          </Stack>
+        ) : (
+          <></>
+        )}
+      </Container>
+    </Box>
+  );
+};
+
+const getCharityData = (charityId: number | undefined): Promise<CharityData | undefined> => {
+  if (charityId === undefined) {
+    // Router is still loading query.
+    return Promise.resolve(undefined);
+  }
+
+  return api.charities.getCharity(charityId).then((res) => {
+    if (res.payload === null) {
+      // Defensive programming to ensure only non-null charity data is populated.
+      throw new Error();
```

Comment:
Instead of throwing a general error can we use these cases?
1. undefined param --> loading
2. undefined payload --> loading
3. null --> failed to fetch intended payload, display error
4. intended payload --> display happy path

This way we wouldn't have to write this function for every swr case, and there's no general error floating around. 

We also don't have to propagate the knowledge of undefined id down to swr call. Suppose we already know id is undefined, we immediately show loading, instead of still passing it to swr which makes it hard to tell if the undefined is due to swr not working or if it's due to us not passing the query param properly.

Reply:
Sgtm!